Client/core/Graphics/CGraphics.cpp

```diff
@@ -795,6 +795,105 @@ void CGraphics::DrawRectQueued(float fX, float fY, float fWidth, float fHeight,
+{
+    if (g_pCore->IsWindowMinimized())
+        return;
+
+    BeginDrawBatch();
+    CheckModes(EDrawMode::DX_SPRITE, m_ActiveBlendMode);
+
+    // Set up a queue item
+    sDrawQueueItem Item;
+    Item.eType = QUEUE_CIRCLE;
+    Item.blendMode = m_ActiveBlendMode;
+    Item.Circle.fX = fX;
+    Item.Circle.fY = fY;
+    Item.Circle.fRadius = fRadius;
+    Item.Circle.startAngle = startAngle;
+    Item.Circle.stopAngle = stopAngle;
+    Item.Circle.bPostGUI = bPostGUI;
+    Item.Circle.fSegments = fSegments;
+    Item.Circle.fRatio = fRatio;
+    Item.Circle.ulColor = color;
+    Item.Circle.ulColorCenter = colorCenter;
+    Item.Circle.bSubPixelPositioning = bSubPixelPositioning;
+    // Add it to the queue
+    AddQueueItem(Item, bPostGUI);
+
+    DrawCircleInternal(fX, fY, fRadius, startAngle, stopAngle, color, colorCenter, fSegments, fRatio, bPostGUI, bSubPixelPositioning);
+
+    EndDrawBatch();
+}
+
+
+struct stVertex
+{
+    float x, y, z;
+    D3DCOLOR color;
+};
+
+void CGraphics::DrawCircleInternal(float fX, float fY, float fRadius, float startAngle, float stopAngle, unsigned long color, unsigned long colorCenter, ushort fSegments, float fRatio, bool bPostGUI, bool bSubPixelPositioning)
+{
+
+    // Adjust size to account for sub pixel borders
+    if (bSubPixelPositioning)
+    {
+        fX += 0.5f;
+        fY += 0.5f;
+    }
+
+    startAngle = D3DXToRadian(startAngle);
+    stopAngle = D3DXToRadian(stopAngle);
+
+    std::vector<stVertex> Points;
+
+    // center
+    stVertex vertCenter;
+    vertCenter.x = fX;
+    vertCenter.y = fY;
+    vertCenter.z = 0;
+    vertCenter.color = colorCenter;
+    Points.push_back(vertCenter);
+    
+    // first
+    stVertex vertFirst;
+    vertFirst.x = fX + fRadius * cos(startAngle) * fRatio;
+    vertFirst.y = fY + fRadius * sin(startAngle) / fRatio;
+    vertFirst.z = 0;
+    vertFirst.color = color;
+    Points.push_back(vertFirst);
+
+    bool out = false;
+    float fSegments2 = (float)fSegments + 1;
+    float segmentAngle = ((stopAngle - startAngle) / fSegments2);
+
+    for (float angle = startAngle; angle <= stopAngle;)
+    {
+        stVertex vertex;
+        vertex.x = fX + fRadius * cos(angle) * fRatio;
+        vertex.y = fY + fRadius * sin(angle) / fRatio;
+        vertex.z = 0;
+        vertex.color = color;
+        Points.push_back(vertex);
+        angle = angle + segmentAngle;
+    }
+
+    // last
+    stVertex vertLast;
+    vertLast.x = fX + fRadius * cos(stopAngle) * fRatio;
+    vertLast.y = fY + fRadius * sin(stopAngle) / fRatio;
+    vertLast.z = 0;
+    vertLast.color = color;
+    Points.push_back(vertLast);
+
+    if (Points.size() >= 3)
+    {
+        m_pDevice->SetTexture(0, 0);
+        m_pDevice->DrawPrimitiveUP(D3DPT_TRIANGLEFAN, Points.size() - 2, &Points[0], sizeof(stVertex));

```

Comment:
Made some researches and it looks like we should avoid using DrawPrimitiveUP at all cost as it is slow.
Source: https://www.gamedev.net/forums/topic/99121-drawprimitiveup-vs-drawprimitive/

I didn't look what should be done to not use the UP version.

Maybe someone who already worked with that stuff could add his input ? I'm just putting a warning flag here.

Reply:
DrawPrimitive would be a better choice if the vertex data remained unchanged between frames. The dynamic nature of dxDraw functions means DrawPrimitiveUP is fine